server/events.go

```diff
@@ -653,6 +654,24 @@ func (s *Server) updateRemoteServer(ms *ServerInfo) {
 	}
 }
 
+// processNewServer will hold any logic we want to use when we discover a new server.
+// Lock should be held upon entry.
+func (s *Server) processNewServer(ms *ServerInfo) {
+	// Right now we will check if we have leafnode servers and if so send another
+	// connect update to make sure they switch this account to interest only mode.
+	// TODO(dlc) - this will cause this account to be loaded on all servers. Need a better
+	// way with GW2.
+	if s.gateway.enabled && len(s.leafs) > 0 {
```

Comment:
With this approach, we are at the mercy of the system event `$SYS.SERVER.ACCOUNT.%s.CONNS` to be sent before switching to interest-only mode in the situation where B is restarted in the setup: `A <-(GW)- B <- (LN) - C`.
This is because C will likely recreate the LN connection to B before B has setup the GW connection to A, so the LN connect event is "lost". Once the GW connection is established, we have to wait for the event hb interval to have A sends the account's conn update for the switch to be made.

I would suggest moving this code to a separate function (and invoke that function from here) as such:
```
// If GW is enabled on this server and there are any leaf node connections,
// this function will send a LeafNode connect system event to the super cluster
// to ensure that the GWs are in interest-only mode for this account.
// Lock should be held upon entry.
func (s *Server) ensureGWsInterestOnlyForLeafNodes() {
	if !s.gateway.enabled || len(s.leafs) == 0 {
		return
	}
	sent := make(map[*Account]bool, len(s.leafs))
	for _, c := range s.leafs {
		if !sent[c.acc] {
			fmt.Printf("@@IK: HERE!\n")
			s.sendLeafNodeConnectMsg(c.acc.Name)
			sent[c.acc] = true
		}
	}
}
```
and call this function in gateway.go:1080, when processing inbound first INFO protocol:
```
    s.mu.Lock()
    s.ensureGWsInterestOnlyForLeafNodes()
    s.mu.Unlock()
```
This will make the switch occur when the GW connection are made, which should also solve the case when there is a disconnect/reconnect of GW connection due to a slow consumer for instance.

Reply:
It's actually not the conns event that triggers it but seeing new servers that will eventually fix it, but most likely they miss the original and need to wait on HB intervals to see the new servers and respond. That is the new code I added.

Reply:
I agree with your suggestion above. It is a blunt instrument for now until we have something like GW2.

Reply:
ok take another look, posted updates.